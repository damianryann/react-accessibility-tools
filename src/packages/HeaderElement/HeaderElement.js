import React from "react";
import PropTypes from "prop-types";

function HeaderElement(props) {
  const { title, isFunctional, value, fallBack, className, id, style } = props;

  const headerOptions = ["h1", "h2", "h3", "h4", "h5", "h6", "p"];

  if (headerOptions.indexOf(value || fallBack) === -1) {
    console.error(
      "Error: value or fallBack must contain the correct header value. Please refer to documentation."
    );
    return (
      <h1>
        Error: <code>value</code> / <code>fallBack</code> must contain the
        correct header value. Please refer to documentation.
      </h1>
    );
  }

  const HeaderElement = value ? value : fallBack;

  function addClassToDiv() {
    let classes = [HeaderElement];

    if (className) {
      classes.push(className);
    }

    return classes.join(" ");
  }

  function addClassToTag() {
    let tagClasses = (document.getElementsByTagName(
      HeaderElement
    ).className = className);

    return tagClasses;
  }

  function addId() {
    let tagIdToTag = (document.getElementsByTagName(HeaderElement).id = id);
    let tagIdToClass = (document.getElementsByClassName(HeaderElement).id = id);

    return tagIdToTag || tagIdToClass;
  }

  function addStyle() {
    let tagIdToTag = (document.getElementsByTagName(
      HeaderElement
    ).style = style);
    let tagIdToClass = (document.getElementsByClassName(
      HeaderElement
    ).style = style);

    return tagIdToTag || tagIdToClass;
  }

  const headerElementMarkup = isFunctional ? (
    <HeaderElement className={addClassToTag()} id={addId()} style={addStyle()}>
      {title}
    </HeaderElement>
  ) : (
    <div className={addClassToDiv()} id={addId()} style={addStyle()}>
      {title}
    </div>
  );

  return headerElementMarkup;
}

HeaderElement.propTypes = {
  title: PropTypes.string,
  isFunctional: PropTypes.bool,
  value: PropTypes.string,
  fallBack: PropTypes.string,
  className: PropTypes.string,
  id: PropTypes.string,
  style: PropTypes.string,
};

export default HeaderElement;
