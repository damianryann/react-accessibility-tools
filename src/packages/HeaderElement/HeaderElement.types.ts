import React from "react";

export interface HeaderProps {
  title: string;
  isFunctional: boolean;
  value: string;
  fallBack?: string;
  className?: string;
  id?: string;
  style?: React.CSSProperties;
}
