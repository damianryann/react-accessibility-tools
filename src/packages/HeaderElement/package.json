{
  "name": "@damianryann/header-element",
  "version": "1.0.1",
  "main": "HeaderElement.tsx",
  "author": "Damian Ryan <damian@damianryan.co.uk>",
  "license": "MIT"
}