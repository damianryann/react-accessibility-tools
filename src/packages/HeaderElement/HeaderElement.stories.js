import React from "react";

import HeaderElement from "./HeaderElement";
import "../../sass/index.scss";

export default { title: "HeaderElement" };

export const functionalHeader1 = () => (
  <HeaderElement isFunctional={true} value="h1" title="Functional Header 1" />
);
export const functionalHeader2 = () => (
  <HeaderElement isFunctional={true} value="h2" title="Functional Header 2" />
);
export const functionalHeader3 = () => (
  <HeaderElement isFunctional={true} value="h3" title="Functional Header 3" />
);
export const functionalHeader4 = () => (
  <HeaderElement isFunctional={true} value="h4" title="Functional Header 4" />
);
export const functionalHeader5 = () => (
  <HeaderElement isFunctional={true} value="h5" title="Functional Header 5" />
);
export const functionalHeader6 = () => (
  <HeaderElement isFunctional={true} value="h6" title="Functional Header 6" />
);
export const decorativeHeader1 = () => (
  <HeaderElement isFunctional={false} value="h1" title="Decorative Header 1" />
);
export const decorativeHeader2 = () => (
  <HeaderElement isFunctional={false} value="h2" title="Decorative Header 2" />
);
export const decorativeHeader3 = () => (
  <HeaderElement isFunctional={false} value="h3" title="Decorative Header 3" />
);
export const decorativeHeader4 = () => (
  <HeaderElement isFunctional={false} value="h4" title="Decorative Header 4" />
);
export const decorativeHeader5 = () => (
  <HeaderElement isFunctional={false} value="h5" title="Decorative Header 5" />
);
export const decorativeHeader6 = () => (
  <HeaderElement isFunctional={false} value="h6" title="Decorative Header 6" />
);
export const functional1WithClass = () => (
  <HeaderElement
    isFunctional={true}
    value="h1"
    title="Funcional Header 1 with 'text-center' class"
    className="text-center"
  />
);
export const decorative1WithClass = () => (
  <HeaderElement
    isFunctional={false}
    value="h1"
    title="Decorative Header 1 with 'text-center' class"
    className="text-center"
  />
);
export const functional1WithStyle = () => (
  <HeaderElement
    isFunctional={true}
    value="h1"
    title="Functional Header 1 with red inline color"
    style={{ color: "#EE4622" }}
  />
);
export const decorative1WithStyle = () => (
  <HeaderElement
    isFunctional={false}
    value="h1"
    title="Decoraive Header 1 with blue inline color"
    style={{ color: "#2159CF" }}
  />
);
