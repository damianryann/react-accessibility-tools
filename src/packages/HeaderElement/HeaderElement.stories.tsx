import React from 'react';

import HeaderElement from './HeaderElement';
import { HeaderProps } from './HeaderElement.types';

export default {
  title: 'React Accessible Components/HeaderElement',
  component: HeaderElement,
  argTypes: {
    heading: {
      control: {
        options: ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p'],
        type: 'select'
      }
    }
  }
};

export const Default = ({ heading, children, ...props }: HeaderProps) => {
  return (
    <HeaderElement heading={heading} {...props}>
      {children}
    </HeaderElement>
  );
};

Default.args = {
  heading: 'h1',
  children: 'Accessible Heading'
};
