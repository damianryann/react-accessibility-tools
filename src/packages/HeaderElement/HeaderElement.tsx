import React, { ElementType, FunctionComponent } from 'react';

import { HeaderProps } from './HeaderElement.types';

export const headerOptions = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p'];

const HeaderElement: FunctionComponent<HeaderProps> = ({
  title,
  isFunctional,
  value,
  fallBack,
  className,
  id,
  style
}) => {
  if (
    !headerOptions.includes(String(value).toLowerCase()) &&
    !headerOptions.includes(String(fallBack).toLowerCase())
  ) {
    const errMessage =
      'Error 001: value or fallBack must contain the correct header value. Please refer to documentation.';
    console.error(errMessage);
    return <code>{errMessage}</code>;
  }

  const divClassHanlder = () => {
    let classes = [
      value.toLowerCase() ?? undefined,
      !value ? fallBack?.toLowerCase() : undefined,
      className ?? undefined
    ].join(' ');

    const removeExtraSpaces = / +/g;

    if (!className) {
      return classes.toString().replaceAll(removeExtraSpaces, '');
    } else {
      return classes.toString().replaceAll(removeExtraSpaces, ' ');
    }
  };

  const Header = (
    value ? value.toLowerCase() : fallBack?.toLowerCase()
  ) as ElementType;

  const headerElementMarkup = isFunctional ? (
    <Header
      className={className ?? undefined}
      id={id ?? undefined}
      style={style ?? undefined}
    >
      {title}
    </Header>
  ) : (
    <div
      className={divClassHanlder() ?? undefined}
      id={id ?? undefined}
      style={style ?? undefined}
    >
      {title}
    </div>
  );

  return headerElementMarkup;
};

export default HeaderElement;
