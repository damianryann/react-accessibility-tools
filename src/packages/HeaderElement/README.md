### HeaderElement

**Available Props:**

| Name           | Type      | Description                                                                       |
| -------------- | --------- | --------------------------------------------------------------------------------- |
| `isFunctional` | `Boolean` | **Required.** Is the element a header or a div class.                             |
| `value`        | `String`  | **Required.** What header is it? h1 - h6 / p.                                     |
| `fallBack`     | `String`  | **Required.** The header you wish to fall back on should no value exist (h1 - h6) |
| `title`        | `String`  | **Required.** The title of your header.                                           |
| `className`    | `String`  | Add additional classes for styling.                                               |
| `id`           | `String`  | Add an ID for behaviours such as anchors.                                         |
| `style`        | `Object`  | Use inline styling should you need to dynamically add CSS.                        |

```js
import React from "react";
import HeaderElement from "react-accessibility-tools";

const ComponentName = () => {
  return (
    <div className="container">
      <div className="row">
        <div className="col-md-12">
          <HeaderElement
            isFunctional={true}
            title="Hello World"
            value="h1"
            fallBack="h2"
          />
        </div>
      </div>
    </div>
  );
};

export default ComponentName;
```

```js
import React from "react";
import HeaderElement from "react-accessibility-tools";

const StandardText = (props) => {
  const { data = {} } = props;
  const { isFuncitonal, title, titleHeader, fallBack } = data;

  return (
    <div className="container">
      <div className="row">
        <div className="col-md-12">
          <HeaderElement
            isFunctional={isFuncitonal}
            title={title}
            value={titleHeader}
            fallBack={fallBack}
          />
        </div>
      </div>
    </div>
  );
};

export default StandardText;
```
