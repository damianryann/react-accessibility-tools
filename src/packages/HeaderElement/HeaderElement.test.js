import React from "react";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { mount } from "enzyme";

import HeaderElement from "./HeaderElement";

Enzyme.configure({ adapter: new Adapter() });

it("mounts without crashing", () => {
  const component = mount(<HeaderElement value="h2" />);

  expect(component).toMatchSnapshot();
});

it("output is a functional header 1 element", () => {
  const component = mount(
    <HeaderElement isFunctional={true} value="h1" title="Hello World" />
  );

  expect(component.find("h1")).toHaveLength(1);
  expect(component.find("h1").html()).toBe(`<h1>Hello World</h1>`);

  expect(component).toMatchSnapshot();
});

it("output is a decorational header 1 element", () => {
  const component = mount(
    <HeaderElement isFunctional={false} value="h1" title="Hello World" />
  );

  expect(component.find(".h1")).toHaveLength(1);
  expect(component.find(".h1").html()).toBe(
    `<div class="h1">Hello World</div>`
  );

  expect(component).toMatchSnapshot();
});

it("fallback takes over if value null", () => {
  const component = mount(
    <HeaderElement isFunctional={true} fallBack="h2" title="Hello World" />
  );

  expect(component.find("h2")).toHaveLength(1);
  expect(component.find("h2").html()).toBe(`<h2>Hello World</h2>`);

  expect(component).toMatchSnapshot();
});

it("classname and id is added to functional element", () => {
  const component = mount(
    <HeaderElement
      isFunctional={true}
      value="h2"
      fallBack="h1"
      title="Hello World"
      className="text-center"
      id="abc-123"
    />
  );

  expect(component.find("h2").hasClass("text-center")).toBeTruthy();
  expect(component.find("#abc-123")).toBeTruthy();

  expect(component).toMatchSnapshot();
});

it("classname and id is added to decorative element", () => {
  const component = mount(
    <HeaderElement
      isFunctional={false}
      value="h2"
      fallBack="h1"
      title="Hello World"
      className="text-center"
      id="abc-123"
    />
  );

  expect(component.find("div").hasClass("text-center")).toBeTruthy();
  expect(component.find("#abc-123")).toBeTruthy();

  expect(component).toMatchSnapshot();
});
