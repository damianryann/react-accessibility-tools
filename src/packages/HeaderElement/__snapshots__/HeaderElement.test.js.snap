// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classname and id is added to decorative element 1`] = `ReactWrapper {}`;

exports[`classname and id is added to functional element 1`] = `ReactWrapper {}`;

exports[`fallback takes over if value null 1`] = `ReactWrapper {}`;

exports[`mounts without crashing 1`] = `ReactWrapper {}`;

exports[`output is a decorational header 1 element 1`] = `ReactWrapper {}`;

exports[`output is a functional header 1 element 1`] = `ReactWrapper {}`;
